nixos/modules/misc/version.nix

```
@@ -121,22 +121,29 @@ in
       default = cfg.release;
       defaultText = literalExpression "config.${opt.release}";
       description = lib.mdDoc ''
-        Every once in a while, a new NixOS release may change
-        configuration defaults in a way incompatible with stateful
-        data. For instance, if the default version of PostgreSQL
-        changes, the new version will probably be unable to read your
-        existing databases. To prevent such breakage, you should set the
-        value of this option to the NixOS release with which you want
-        to be compatible. The effect is that NixOS will use
-        defaults corresponding to the specified release (such as using
-        an older version of PostgreSQL).
-        It’s perfectly fine and recommended to leave this value at the
-        release version of the first install of this system.
-        Changing this option will not upgrade your system. In fact it
-        is meant to stay constant exactly when you upgrade your system.
-        You should only bump this option, if you are sure that you can
-        or have migrated all state on your system which is affected
-        by this option.
+        This option defines the first version of NixOS you have installed on this particular machine,
+        and is used to maintain compatibility with data not managed by NixOS configuration (e.g. databases)
+        that was created by applications running on older NixOS versions.
+
+        For example, if NixOS version XX.YY ships with AwesomeDB version N by default, and is then
+        upgraded to version XX.YY+1, which ships AwesomeDB version N+1, the existing databases
+        may no longer be compatible, causing applications to fail, or even data loss.
+
+        The `stateVersion` mechanism avoids this situation by making the default version of such packages
+        conditional on the first version of NixOS you've installed (encoded in `stateVersion`), instead of
+        simply always using the latest one.
+
+        Note that this generally only affects applications that can't upgrade their data automatically -
+        applications and services supporting automatic migrations will remain on latest versions when
+        you upgrade.
+
+        Most users should **never** change this value after the initial install, for any reason.
+
+        This value does **not** affect the Nixpkgs version your packages and OS are pulled from,
+        so changing it will **not** upgrade your system.
```

```suggestion
        so changing it will **not** upgrade your system. Conversely, leaving it unchanged it does **not** mean your system is out of date. This value being older than the release of NixOS you are currently using is normal and expected.
```
Same reasoning as above.

---

Huh, it came out above the other one on the main pr page... oh well, "below", then.

---

This feels a bit too long for me. I really tried to keep these lines as short and punchy as possible.

---

I think it might be more important to say that leaving it unchanged doesn't mean you're out of date, than to say that changing it doesn't upgrade the system. They're essentially the same thing, but what motivates people to change it is usually a fear that they'll somehow be out of date if they don't.

---

I guess we can add it as another bullet point.

---

Found a good wording, I think.